File: clients/ios/StellarChat/StellarChat/Models/StellarStrKey.swift
import Foundation

/// Stellar StrKey encoding (SEP-0023).
/// Encodes Ed25519 public keys as G..., secret keys as S...
enum StellarStrKey {
    /// Version byte for Ed25519 public key (account ID): 6 << 3 = 48.
    private static let versionAccountID: UInt8 = 6 << 3

    /// Encode a 32-byte Ed25519 public key as a Stellar account ID (G...).
    static func encodeAccountID(_ publicKey: Data) -> String {
        precondition(publicKey.count == 32, "Ed25519 public key must be 32 bytes")
        var payload = Data([versionAccountID])
        payload.append(publicKey)
        let checksum = crc16XModem(payload)
        payload.append(checksum.littleEndianBytes)
        return base32Encode(payload)
    }

    // MARK: - CRC16-XModem

    private static func crc16XModem(_ data: Data) -> UInt16 {
        var crc: UInt16 = 0x0000
        for byte in data {
            crc ^= UInt16(byte) << 8
            for _ in 0..<8 {
                if crc & 0x8000 != 0 {
                    crc = (crc << 1) ^ 0x1021
                } else {
                    crc <<= 1
                }
            }
        }
        return crc & 0xFFFF
    }

    // MARK: - Base32 (RFC 4648, no padding)

    private static let base32Alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567")

    private static func base32Encode(_ data: Data) -> String {
        var result = ""
        result.reserveCapacity((data.count * 8 + 4) / 5)

        var buffer: UInt64 = 0
        var bitsLeft = 0

        for byte in data {
            buffer = (buffer << 8) | UInt64(byte)
            bitsLeft += 8
            while bitsLeft >= 5 {
                bitsLeft -= 5
                let index = Int((buffer >> bitsLeft) & 0x1F)
                result.append(base32Alphabet[index])
            }
        }

        if bitsLeft > 0 {
            let index = Int((buffer << (5 - bitsLeft)) & 0x1F)
            result.append(base32Alphabet[index])
        }

        return result
    }
}

private extension UInt16 {
    var littleEndianBytes: Data {
        var value = self.littleEndian
        return withUnsafeBytes(of: &value) { Data($0) }
    }
}

File: clients/android/StellarChat/app/src/main/java/chat/onym/android/crypto/StellarStrKey.kt
package chat.onym.android.crypto

/**
 * Stellar StrKey encoding (SEP-0023).
 * Encodes Ed25519 public keys as "G..." account IDs.
 */
object StellarStrKey {
    private const val VERSION_ACCOUNT_ID: Byte = (6 shl 3).toByte() // 48

    private val BASE32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567".toCharArray()

    /** Encode a 32-byte Ed25519 public key as a Stellar account ID (G...). */
    fun encodeAccountID(publicKey: ByteArray): String {
        require(publicKey.size == 32) { "Ed25519 public key must be 32 bytes" }

        val payload = ByteArray(35) // 1 version + 32 key + 2 checksum
        payload[0] = VERSION_ACCOUNT_ID
        System.arraycopy(publicKey, 0, payload, 1, 32)

        val checksum = crc16XModem(payload, 0, 33)
        payload[33] = (checksum and 0xFF).toByte()
        payload[34] = ((checksum shr 8) and 0xFF).toByte()

        return base32Encode(payload)
    }

    /** CRC-16-XModem (polynomial 0x1021, initial value 0x0000). */
    private fun crc16XModem(data: ByteArray, offset: Int, length: Int): Int {
        var crc = 0x0000
        for (i in offset until offset + length) {
            crc = crc xor ((data[i].toInt() and 0xFF) shl 8)
            for (j in 0 until 8) {
                crc = if (crc and 0x8000 != 0) {
                    (crc shl 1) xor 0x1021
                } else {
                    crc shl 1
                }
                crc = crc and 0xFFFF
            }
        }
        return crc
    }

    /** RFC 4648 Base32 encoding (no padding). */
    private fun base32Encode(data: ByteArray): String {
        val sb = StringBuilder()
        var buffer = 0
        var bitsLeft = 0

        for (byte in data) {
            buffer = (buffer shl 8) or (byte.toInt() and 0xFF)
            bitsLeft += 8
            while (bitsLeft >= 5) {
                bitsLeft -= 5
                sb.append(BASE32_ALPHABET[(buffer shr bitsLeft) and 0x1F])
            }
        }
        if (bitsLeft > 0) {
            sb.append(BASE32_ALPHABET[(buffer shl (5 - bitsLeft)) and 0x1F])
        }
        return sb.toString()
    }
}
